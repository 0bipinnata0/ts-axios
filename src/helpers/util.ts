const toString = Object.prototype.toString

export function isDate(val: any): val is Date {
  return toString.call(val) === '[object Date]'
}

export function isPlainObject(val: any): val is Object {
  return toString.call(val) === '[object Object]'
}

export function isFalsy(val: any): boolean {
  return val === undefined || val === null
}

export function isArray<T>(val: any): val is Array<T> {
  return Array.isArray(val)
}

export function isString(val: any): val is string {
  return typeof val === 'string'
}
