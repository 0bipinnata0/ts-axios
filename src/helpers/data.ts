import { isPlainObject } from './util'

export function transformRequest(data: unknown) {
  if (isPlainObject(data)) {
    return JSON.stringify(data)
  }
  return data
}
