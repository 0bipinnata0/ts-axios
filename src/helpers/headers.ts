import { isPlainObject } from './util'

function normalizeHeaderName(headers: any, normalizedName: string): void {
  if (!headers) {
    return
  }

  Object.keys(headers).forEach((name) => {
    if (name !== normalizedName && name.toUpperCase() === normalizedName.toUpperCase()) {
      headers[normalizedName] = headers[name]
      delete headers[name]
    }
  })
}

export function processHeaders(headers: any, data: any): any {
  normalizeHeaderName(headers, 'Content-Type')

  if (isPlainObject(data)) {
    if (headers && !headers['Content-Type']) {
      headers['Content-Type'] = 'application/json;charset=utf-8'
    }
  }
  return headers
}

export function parseHeaders(headers: string): Record<string, any> {
  let parsed = Object.create(null)
  if (!headers) {
    return parsed
  }
  const entities = headers
    .split('\r\n')
    .map((i) => i.trim())
    // 清楚无效项
    .filter((i) => i.length > 0)
    // 清楚无效项
    .filter((i) => i[0] !== ':')
    .map((line) => {
      const [key, ...values] = line.split(':')
      return [key, values.join(':').trim()]
    })

  return Object.fromEntries(entities)
}
