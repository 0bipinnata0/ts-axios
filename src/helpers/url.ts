import { isArray, isDate, isFalsy, isPlainObject } from './util'

function encode(val: string): string {
  return encodeURIComponent(val)
    .replace(/%40/g, '@')
    .replace(/%3A/gi, ':')
    .replace(/%24/g, '$')
    .replace(/%2C/gi, ',')
    .replace(/%20/g, '+')
    .replace(/%5B/gi, '[')
    .replace(/%5D/gi, ']')
}

export function buildURL(url: string, params?: any): string {
  if (!params) {
    return url
  }

  {
    const serializedParams = Object.entries(params)
      .filter(([, val]) => !isFalsy(val)) // 忽略空值
      .map(([key, val]) => (isArray(val) ? [key + '[]', val] : [key, val])) // 处理数组
      .map(([key, value]) => (isDate(value) ? [key, value.toISOString()] : [key, value])) // 处理日期
      .map(([key, value]) => (isPlainObject(value) ? [key, JSON.stringify(value)] : [key, value])) // 处理对象
      .map(([key, value]) => [encode(`${key}`), encode(`${value}`)]) // 编码
      .join('&') // 拼接

    const markIndex = url.indexOf('#')
    const connectIndex = url.indexOf('?')
    const parsedUrl = url.slice(0, markIndex) + (connectIndex === -1 ? '?' : '&') + serializedParams
    console.info('parsedUrl', parsedUrl)
  }

  const parts: string[] = []

  Object.keys(params).forEach((key) => {
    const val = params[key]
    if (isFalsy(val)) {
      return
    }

    let values = []
    if (Array.isArray(val)) {
      values = val
      key += '[]'
    } else {
      values = [val]
    }

    values.forEach((val) => {
      if (isDate(val)) {
        val = val.toISOString()
      } else if (isPlainObject(val)) {
        val = JSON.stringify(val)
      }
      parts.push(`${encode(key)}=${encode(val)}`)
    })
  })

  let serializedParams = parts.join('&')

  if (serializedParams) {
    const markIndex = url.indexOf('#')
    if (markIndex !== -1) {
      url = url.slice(0, markIndex)
    }

    url += (url.indexOf('?') === -1 ? '?' : '&') + serializedParams
  }

  return url
}
