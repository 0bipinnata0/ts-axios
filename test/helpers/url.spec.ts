import { buildURL } from '../../src/helpers/url'

describe('helpers: url', () => {
  describe('buildURL', () => {
    test('buildURL 基本使用', () => {
      expect(
        buildURL('/simple/get', {
          method: 'get',
          params: {
            a: 1,
            b: 2,
          },
        })
      ).toMatchSnapshot()
    })

    test('buildURL 参数值为数组', () => {
      expect(
        buildURL('/simple/get', {
          method: 'get',
          params: {
            foo: ['bar', 'baz'],
          },
        })
      ).toMatchSnapshot()
    })

    test('buildURL 参数值为对象', () => {
      expect(
        buildURL('/simple/get', {
          method: 'get',
          params: {
            foo: {
              bar: 'baz',
            },
          },
        })
      ).toMatchSnapshot()
    })

    test('buildURL 空值(undefined | null)', () => {
      expect(
        buildURL('/simple/get', {
          method: 'get',
          params: {
            foo: {
              foo: 'bar',
              baz: null, // 被忽略
            },
          },
        })
      ).toMatchSnapshot()
    })

    test('buildURL 丢弃 url 中的哈希标记', () => {
      expect(
        buildURL('/simple/get#hash', {
          method: 'get',
          params: {
            foo: {
              foo: 'bar',
              baz: null, // 被忽略
            },
          },
        })
      ).toMatchSnapshot()
    })

    test('buildURL 特殊字符支持', () => {
      expect(
        buildURL('/simple/get#hash', {
          method: 'get',
          params: {
            foo: {
              foo: '@:$, ',
            },
          },
        })
      ).toMatchSnapshot()
    })

    test('buildURL 保留 url 中已经存在的参数', () => {
      expect(
        buildURL('/base/get?boo=bar', {
          method: 'get',
          params: {
            foo: {
              foo: '@:$, ',
            },
          },
        })
      ).toMatchSnapshot()
    })

    test('buildURL参数值为 Date', () => {
      const date = new Date(1679758711189)
      expect(
        buildURL('/base/get?boo=bar', {
          method: 'get',
          params: {
            foo: {
              date,
            },
          },
        })
      ).toMatchSnapshot()
    })
  })
})
