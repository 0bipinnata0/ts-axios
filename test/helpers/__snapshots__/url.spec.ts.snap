// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers: url buildURL buildURL 丢弃 url 中的哈希标记 1`] = `"/simple/get?method=get&params=%7B%22foo%22:%7B%22foo%22:%22bar%22,%22baz%22:null%7D%7D"`;

exports[`helpers: url buildURL buildURL 保留 url 中已经存在的参数 1`] = `"/base/get?boo=bar&method=get&params=%7B%22foo%22:%7B%22foo%22:%22@:$,+%22%7D%7D"`;

exports[`helpers: url buildURL buildURL 参数值为对象 1`] = `"/simple/get?method=get&params=%7B%22foo%22:%7B%22bar%22:%22baz%22%7D%7D"`;

exports[`helpers: url buildURL buildURL 参数值为数组 1`] = `"/simple/get?method=get&params=%7B%22foo%22:[%22bar%22,%22baz%22]%7D"`;

exports[`helpers: url buildURL buildURL 基本使用 1`] = `"/simple/get?method=get&params=%7B%22a%22:1,%22b%22:2%7D"`;

exports[`helpers: url buildURL buildURL 特殊字符支持 1`] = `"/simple/get?method=get&params=%7B%22foo%22:%7B%22foo%22:%22@:$,+%22%7D%7D"`;

exports[`helpers: url buildURL buildURL 空值(undefined | null) 1`] = `"/simple/get?method=get&params=%7B%22foo%22:%7B%22foo%22:%22bar%22,%22baz%22:null%7D%7D"`;

exports[`helpers: url buildURL buildURL参数值为 Date 1`] = `"/base/get?boo=bar&method=get&params=%7B%22foo%22:%7B%22date%22:%222023-03-25T15:38:31.189Z%22%7D%7D"`;
